File: 2-trap/trap-mpi.f90
MODULE precision
  INTEGER ,PARAMETER:: long=SELECTED_REAL_KIND(15,307) !(6,37) for real(float), (15,307) for double
END MODULE precision

module time_ftcs
    contains
      function timestamp() result(str)
        implicit none
        character(len=20)                     :: str
        integer                               :: values(8)
        character(len=4)                      :: year
        character(len=2)                      :: month
        character(len=2)                      :: day, hour, minute, second
        character(len=5)                      :: zone
    
        ! Get current time
        call date_and_time(VALUES=values, ZONE=zone)  
        write(year,'(i4.4)')    values(1)
        write(month,'(i2.2)')   values(2)
        write(day,'(i2.2)')     values(3)
        write(hour,'(i2.2)')    values(5)
        write(minute,'(i2.2)')  values(6)
        write(second,'(i2.2)')  values(7)
    
        str = day//'-'//hour//'-'//minute//'-'//second
      end function timestamp
end module


program main
    use precision
    use time_ftcs
    use iso_fortran_env
    implicit none
    include 'mpif.h'
    
    ! Calculate local integral
    INTERFACE 

       SUBROUTINE Trap(left_endpt, right_endpt, trap_count, base_len, estimate)
            use precision
            use iso_fortran_env
            IMPLICIT NONE
            REAL(long), INTENT(IN) :: left_endpt
            REAL(long), INTENT(IN) :: right_endpt
            REAL(long), INTENT(IN) :: base_len
            INTEGER, INTENT(IN) :: trap_count
            REAL(long), INTENT(OUT) :: estimate
       END SUBROUTINE Trap

       subroutine OpenFile(a, b, n)
            use precision
            use iso_fortran_env
            implicit none
            REAL(long), INTENT(OUT) :: a
            REAL(long), INTENT(OUT) :: b
            INTEGER, INTENT(OUT) :: n
        end subroutine OpenFile
    END INTERFACE
    
    ! Declaration of variables
    INTEGER :: my_rank 
    INTEGER :: comm_sz
    INTEGER :: n
    INTEGER :: local_n
    INTEGER :: ierror
    INTEGER :: status(MPI_STATUS_SIZE)
    INTEGER :: source
    REAL(long) :: a
    REAL(long) :: b
    REAL(long) :: h
    REAL(long) :: local_a
    REAL(long) :: local_b
    REAL(long) :: local_int
    REAL(long) :: total_int
    REAL(long) :: estimation                        ! Estimacion del calculo
    REAL(long) :: error                             ! Error encontrado
    REAL(long) :: elapsed                           ! Tiempo que tomo ejecutarse el programa
    REAL(long) :: start_clock                       ! Tiempo de inicio
    REAL(long) :: stop_clock                        ! Tiempo de termino
    CHARACTER(len=38) :: start                      ! Hora de inicio
    CHARACTER(len=38) :: end                        ! Hora de termino
    
    ! Let the system do what it needs to start up MPI
    call MPI_INIT(ierror)
    ! Get my process rank
    call MPI_COMM_RANK(MPI_COMM_WORLD, my_rank, ierror)
    ! Find out how many processes are being used
    call MPI_COMM_SIZE(MPI_COMM_WORLD, comm_sz, ierror)
    
    if (my_rank == 0) then

        ! Establece la hora de inicio
        start = timestamp()
        call CPU_TIME(start_clock)

         ! Assignment
        CALL OpenFile(a, b, n)
        
    end if

    call MPI_BCAST(a, 1, MPI_DOUBLE, 0, MPI_COMM_WORLD, ierror)
    call MPI_BCAST(b, 1, MPI_DOUBLE, 0, MPI_COMM_WORLD, ierror)
    call MPI_BCAST(n, 1, MPI_LONG, 0, MPI_COMM_WORLD, ierror)

    h = (b - a) / n         ! h is the same for all processes 
    local_n = n / comm_sz   ! So is the number of trapezoids 
    
    ! Length of each process' interval of
    ! integration = local_n * h.  So my interval
    ! starts at:
    local_a = a + my_rank * local_n * h
    local_b = local_a + local_n * h
    call Trap (local_a, local_b, local_n, h, local_int)

    call MPI_REDUCE(local_int, total_int, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD, ierror)

    ! Print the result
    if (my_rank == 0) then
        
         ! Calculo estimado con la formula a^2*N^3
        estimation = 9.0_long

        ! Calcula el % de error
        error = DABS(total_int - estimation) / estimation * 100.0_long

        end = timestamp()
        call CPU_TIME(stop_clock)
        
        ! Calcula el tiempo que tomo ejecutar el programa
        elapsed =  stop_clock - start_clock

        ! Guarda el resultado en un archivo
        call SaveFile(start, end, error, elapsed, N, comm_sz)

    end if    
    
    call MPI_FINALIZE(ierror)
    
end program main
    
!---------------------------------------------------------------------
!
!  Subroutine for estimating a definite integral 
!  using the trapezoidal rule
! Input args:   left_endpt
!               right_endpt
!               trap_count 
!               base_len
! Output val:   Trapezoidal rule estimate of integral from
!               left_endpt to right_endpt using trap_count
!               trapezoids
!
!---------------------------------------------------------------------
SUBROUTINE Trap(left_endpt, right_endpt, trap_count, base_len, estimate)
    use precision
    use iso_fortran_env
    IMPLICIT NONE
    REAL(long), INTENT(IN) :: left_endpt
    REAL(long), INTENT(IN) :: right_endpt
    REAL(long), INTENT(IN) :: base_len
    INTEGER, INTENT(IN) :: trap_count
    REAL(long), INTENT(OUT) :: estimate
    REAL(long) :: x
    INTEGER :: i
    
    estimate = ( left_endpt ** 2 + right_endpt ** 2 ) / 2_long;
    
    do i = 1, trap_count - 1, 1
    
        x = left_endpt + i * base_len
        estimate = estimate + x ** 2
    
    end do
    
    estimate = estimate * base_len
    
END SUBROUTINE Trap

subroutine OpenFile(a, b, n)
    use precision
    use iso_fortran_env
    implicit none
    REAL(long), INTENT(OUT) :: a
    REAL(long), INTENT(OUT) :: b
    INTEGER, INTENT(OUT) :: n
    INTEGER :: stat
    CHARACTER(len=30) :: input_file

    input_file = 'parameters.dat'

    OPEN(UNIT=1,file=input_file,ACTION="READ",IOSTAT=stat,STATUS='OLD')
    !a = 0.0_long
    !b = 3.0_long
    READ(1, *) a
    READ(1, *) b
    READ(1, *) n

    CLOSE(1)

end subroutine OpenFile


! Crea un archivo de salida con la hora de inicio, de termino y el tiempo que tomo correr el programa
! Asi como el porcentaje de error
subroutine SaveFile(start, end, error, elapsed, N, process)
    use precision
    CHARACTER(len=38), INTENT(IN) :: start
    CHARACTER(len=38), INTENT(IN) :: end
    REAL(long), INTENT(IN) :: error
    REAL(long), INTENT(IN) :: elapsed
    INTEGER, INTENT(IN) :: N
    INTEGER, INTENT(IN) :: process
    CHARACTER(len = 84) :: file_name
    CHARACTER(len = 15) :: Nstring
    CHARACTER(len = 1) :: processString

    write(Nstring, "(I15)") N 
    write(processString, "(I1)") process

    file_name = 'mpi-f90-'//processString//'-'//Nstring//'-'//trim(end)//'.txt'
    
    open (unit=10,file=file_name)
    write(10,*) 'Hora de inicio'
    write(10,*) start
    write(10,*) 'Hora de termino'
    write(10,*) end
    write(10,*) 'Tiempo de ejecucion'
    write(10,*) elapsed
    write(10,*) 'Error'
    write(10,*) error
    close(10)

end subroutine SaveFile

File: 2-trap/trap-mpi.c
/* File:     mpi_trap1.c
 * Purpose:  Use MPI to implement a parallel version of the trapezoidal 
 *           rule.  In this version the endpoints of the interval and
 *           the number of trapezoids are hardwired.
 *
 * Input:    None.
 * Output:   Estimate of the integral from a to b of f(x)
 *           using the trapezoidal rule and n trapezoids.
 *
 * Compile:  mpicc -g -Wall -o mpi_trap1 mpi_trap1.c
 * Run:      mpiexec -n <number of processes> ./mpi_trap1
 *
 * Algorithm:
 *    1.  Each process calculates "its" interval of
 *        integration.
 *    2.  Each process estimates the integral of f(x)
 *        over its interval using the trapezoidal rule.
 *    3a. Each process != 0 sends its integral to 0.
 *    3b. Process 0 sums the calculations received from
 *        the individual processes and prints the result.
 *
 * Note:  f(x), a, b, and n are all hardwired.
 *
 * IPP:   Section 3.2.2 (pp. 96 and ff.)
 */
#include <stdio.h>
#include <math.h>
#include <time.h>

/* We'll be using MPI routines, definitions, etc. */
#include <mpi.h>

/* Calculate local integral  */
double Trap(double left_endpt, double right_endpt, int trap_count, double base_len);
void OpenFile(double *a, double *b, long *n);
void SaveFile(struct tm *start, struct tm *end, double error, double elapsed, long N, int rank);

int main(int argc, char *argv[])
{
   int my_rank;         //
   int comm_sz;         //
   long n;              //
   int local_n;         //
   int source;          //
   double a;            //
   double b;            //
   double h;            //
   double local_a;      //
   double local_b;      //
   double local_int;    //
   double total_int;    //
   double estimation;   // Estimacion del calculo
   double error;        // Error encontrado
   double elapsed;      // Tiempo que tomo ejecutarse el programa
   time_t t;            // Variable de tiempo
   struct tm *start;    // Hora de inicio
   struct tm *end;      // Hora de termino
   clock_t start_clock; // Tiempo de inicio
   clock_t stop_clock;  // Tiempo de termino

   /* Let the system do what it needs to start up MPI */
   MPI_Init(&argc, &argv);

   /* Get my process rank */
   MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

   /* Find out how many processes are being used */
   MPI_Comm_size(MPI_COMM_WORLD, &comm_sz);

   if (my_rank == 0)
   {
      // Inicia la variable de tiempo
      t = time(NULL);

      // Establece la hora de inicio
      start = localtime(&t);
      start_clock = clock();

      OpenFile(&a, &b, &n);
   }

   MPI_Bcast(&a, 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);
   MPI_Bcast(&b, 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);
   MPI_Bcast(&n, 1, MPI_LONG, 0, MPI_COMM_WORLD);

   h = (b - a) / n;       /* h is the same for all processes */
   local_n = n / comm_sz; /* So is the number of trapezoids  */

   /* Length of each process' interval of
    * integration = local_n*h.  So my interval
    * starts at: */
   local_a = a + my_rank * local_n * h;
   local_b = local_a + local_n * h;
   local_int = Trap(local_a, local_b, local_n, h);

   /* Add up the integrals calculated by each process */
   MPI_Reduce(&local_int, &total_int, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

   //
   if (my_rank == 0)
   {
      // Estimacion del resultado
      estimation = 9.0;

      // Calcula el porcentaje de error.
      error = fabs(total_int - estimation) / estimation * 100.0;

      // Establece la hora en que termino de calcular
      end = localtime(&t);
      stop_clock = clock();

      // Calcula el tiempo que tomo ejecutar el programa
      elapsed = (double)(stop_clock - start_clock) / CLOCKS_PER_SEC;

      // Guarda el resultado en un archivo
      SaveFile(start, end, error, elapsed, n, comm_sz);
   }

   /* Shut down MPI */
   MPI_Finalize();

   return 0;
} /*  main  */

/*------------------------------------------------------------------
 * Function:     Trap
 * Purpose:      Serial function for estimating a definite integral 
 *               using the trapezoidal rule
 * Input args:   left_endpt
 *               right_endpt
 *               trap_count 
 *               base_len
 * Return val:   Trapezoidal rule estimate of integral from
 *               left_endpt to right_endpt using trap_count
 *               trapezoids
 */
double Trap(double left_endpt, double right_endpt, int trap_count, double base_len)
{
   double estimate;
   long i;

   estimate = (pow(left_endpt, 2) + pow(right_endpt, 2)) / 2.0;

   for (i = 1; i <= trap_count - 1; i++)
      estimate += pow(left_endpt + i * base_len, 2);

   estimate = estimate * base_len;

   return estimate;
}

void OpenFile(double *a, double *b, long *n)
{
   FILE *file;
   char *input_file;

   input_file = "parameters.dat";

   file = fopen(input_file, "r");

   fscanf(file, "%lf", a);
   fscanf(file, "%lf", b);
   fscanf(file, "%ld", n);

   *n = (long)pow((double)*n, 2);

   fclose(file);
}
// Crea un archivo de salida con la hora de inicio, de termino y el tiempo que tomo correr el programa
// Asi como el porcentaje de error
void SaveFile(
    struct tm *start, // Hora de inicio
    struct tm *end,   // Hora de termino
    double error,     // Porcentaje de error
    double elapsed,   // Tiempo que paso
    long N,           // Dimension de la matriz
    int rank)
{
   FILE *file;
   char file_name[64];
   char output[50];

   sprintf(file_name, "mpi-%d-gcc-%ld-%d-%d-%d-%d.txt", rank, N, end->tm_mday, end->tm_hour, end->tm_min, end->tm_sec);

   file = fopen(file_name, "w+");

   strftime(output, sizeof(output), "%c", start);
   fprintf(file, "Hora de inicio\n%s\n", output);

   strftime(output, sizeof(output), "%c", end);
   fprintf(file, "Hora de termino\n%s\n", output);

   fprintf(file, "Tiempo de ejecucion\n%.15lf\n", elapsed);

   fprintf(file, "Error\n%.15le\n", error);

   fclose(file);
}